File: 2845-count-of-interesting-subarrays/2845-count-of-interesting-subarrays.java
class Solution {
    public long countInterestingSubarrays(List<Integer> nums, int modulo, int k) {
        Map<Long, Long> mp = new HashMap();
        
        long ans =0, prefix=0;
        int n =  nums.size();
        
        mp.put(0L, 1L);
        
        for(int i =0 ; i<n; i++){
            int num = nums.get(i);
            if(num % modulo  == k){
                prefix++;
            }
            
            prefix %= modulo;
            
            long target = (prefix - k + modulo) % modulo;
            ans += mp.getOrDefault(target,0L);
            
            mp.put(prefix, mp.getOrDefault(prefix, 0L) + 1);
        }
        return ans;
    }
}

File: 2845-count-of-interesting-subarrays/2845-count-of-interesting-subarrays.cpp
class Solution {
public:
    long long countInterestingSubarrays(vector<int>& nums, int modulo, int k) {
        unordered_map<long long, long long> mp;
        long long ans =0, prefix=0, n=nums.size();
        
        mp[0]++;
        
        for(int i =0; i<n; i++){
            if(nums[i]%modulo ==k){
                prefix++;
            }
            
            prefix%=modulo;
            
            ans+=mp[(prefix-k+modulo)%modulo];
            
            mp[prefix]++;
            
        }
        return ans;
    }
};